### ISSUE TITLE
Question: How do I model `exists` with `Maybe`?

### ISSUE THREAD

**Issue by ocharles** [LeventErkok/sbv#400]
I'm currently working with `sbv` to solve a classical AI planning problem, but am unsure how to model `Maybe` values. Here's a small example to work with.

We have some sort of workshop with machines, and a way of transporting widgets between machines. Machines can either have a widget in them, or they can be empty. The problem is: given a known initial state, find a series of transitions that leaves in me in some expected state. I'd like to model this with:

```haskell
{-# language TemplateHaskell #-}
{-# language StandaloneDeriving #-}
{-# language DeriveDataTypeable #-}
{-# language DeriveAnyClass #-}

import Data.SBV

type WidgetId = Integer

data State = State 
  { machineA :: SBV (Maybe WidgetId)
  , machineB :: SBV (Maybe WidgetId)
  }

data Action = MoveFromAToB | MoveFromBToA
mkSymbolicEnumeration ''Action

newAction :: Symbolic (SBV Action)
newAction = exists_

-- Given a state, constrain the possible set of actions
constrainAction :: State -> SBV Action -> Symbolic ()
constrainAction s a =
  constrain $
  ite (a .== literal MoveFromAToB)
  (bnot (machineA s .== literal Nothing) &&& machineB s .== literal Nothing) $
  ite (a .== literal MoveFromBToA)
  (bnot (machineB s .== literal Nothing) &&& machineA s .== literal Nothing) $
  true

apply :: State -> SBV Action -> State
apply s a =
  s { machineA =
       ite (a .== literal MoveFromAToB) (literal Nothing) $
       ite (a .== literal MoveFromBToA) (machineB s) $
       (machineA s)
    , machineB =
       ite (a .== literal MoveFromAToB) (machineA s) $
       ite (a .== literal MoveFromBToA) (literal Nothing) $
       (machineA s)
  }

plan :: IO SatResult
plan =
  sat $ do
    let
      s0 = State { machineA = literal Nothing, machineB = literal (Just 5) }

    action <- newAction

    constrainAction s0 action

    let s1 = apply s0 action

    constrain (machineA s1 .== literal (Just 5))
```

Essentially I have some symbolic state, actions that act on state, pre-conditions that constrain actions given a state, and a way to progress the state from an action. I can then solve planning problems using iterative deepening - starting with a known state, can I get to the state I want in 1 step? Want about 2? 3 steps? Etc. This is doable with `query`, `checkSat`, etc.

The problem is... you can't write the above code!

```
    • No instance for (SymWord (Maybe WidgetId))
        arising from a use of ‘ite’
    • In the expression:
        ite (a .== literal MoveFromAToB) (literal Nothing)
      In the ‘machineA’ field of a record
      In the expression:
        s {machineA = ite (a .== literal MoveFromAToB) (literal Nothing)
                        $ ite (a .== literal MoveFromBToA) (machineB s) $ (machineA s),
           machineB = ite (a .== literal MoveFromAToB) (machineA s)
                        $ ite (a .== literal MoveFromBToA) (literal Nothing)
                            $ (machineA s)}
   |
34 |        ite (a .== literal MoveFromAToB) (literal Nothing) $
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

My workaround so far has been to essentially distribute the sum type in `Maybe` into a product type:

```haskell
data State = State 
  { machineA :: (SBV Bool, SBV WidgetId)
  , machineB :: (SBV Bool, SBV WidgetId)
  }

data Action = MoveFromAToB | MoveFromBToA
mkSymbolicEnumeration ''Action

newAction :: Symbolic (SBV Action)
newAction = exists_

-- Given a state, constrain the possible set of actions
constrainAction :: State -> SBV Action -> Symbolic ()
constrainAction s a =
  constrain $
  ite (a .== literal MoveFromAToB)
  (fst (machineA s) &&& bnot (fst (machineB s))) $
  ite (a .== literal MoveFromBToA)
  (bnot (fst (machineA s)) &&& fst (machineB s)) $
  true

apply :: State -> SBV Action -> State
apply s a =
  s { machineA =
       ite (a .== literal MoveFromAToB) (false, 0) $
       ite (a .== literal MoveFromBToA) (machineB s) $
       (machineA s)
    , machineB =
       ite (a .== literal MoveFromAToB) (machineA s) $
       ite (a .== literal MoveFromBToA) (false, 0) $
       (machineA s)
  }

plan :: IO SatResult
plan =
  sat $ do
    let
      s0 = State { machineA = (false, 0), machineB = (true, 5) }

    action <- newAction

    constrainAction s0 action

    let s1 = apply s0 action

    constrain (snd ( machineA s1 ) .== 5)
```

But now for `literal Nothing` I have to write `(false, 0)` - inventing some arbitrary `WidgetId`. Is this the right solution?

I should note that in my actual problem I have a finite static amount of `WidgetId`s. I guess I'd really like to have

```haskell
data MaybeWidgetId = NoWidgetId | Widget1 | Widget2 | Widget3 | ...
```

but I only have this information at runtime.

**Comment by ocharles** [comment]
If this is the wrong place to ask questions I'm happy to re-ask this on Stack Overflow (I see you're quite active there!)

**Comment by LeventErkok** [comment]
Flattening the `Maybe` into a pair of `SBool` and value is a common trick, and is just fine. You do, however, have a typo in your `apply` function, the second to last line should say `machineB s`.

Having said that, I think it might be better to structure so you actually keep track of the current machine and a move is simply the next machine to pass things onto.  Of course, not knowing your exact problem, it's hard to comment. You can probably generalize from this, but I'd be inclined to go with something like:

```haskell
{-# LANGUAGE TemplateHaskell      #-}
{-# LANGUAGE StandaloneDeriving   #-}
{-# LANGUAGE DeriveDataTypeable   #-}
{-# LANGUAGE DeriveAnyClass       #-}
{-# LANGUAGE NamedFieldPuns       #-}
{-# LANGUAGE TypeSynonymInstances #-}
{-# LANGUAGE FlexibleInstances    #-}
{-# LANGUAGE DeriveGeneric        #-}

import Control.Monad.State (State, runState, withState, put, get, execState)
import GHC.Generics (Generic)

import Data.SBV

data Machine = MachineA | MachineB

mkSymbolicEnumeration ''Machine
type SMachine = SBV Machine

data Widget = NoWidget | Widget1 | Widget2 | Widget3 | Widget4 | Widget5

mkSymbolicEnumeration ''Widget
type SWidget = SBV Widget

data Status = Status { machine :: SMachine
                     , widgets :: SFunArray Machine Widget
                     }
                     deriving (Generic, Mergeable)

type Move a = State Status a

instance Mergeable a => Mergeable (Move a) where
  symbolicMerge f t a b = do s <- get
                             let (ar, s1) = runState a s
                                 (br, s2) = runState b s
                             put $ symbolicMerge f t s1 s2
                             return $ symbolicMerge f t ar br

-- An action is simply the next machine to move to
type Action = SMachine

move :: Action -> Status -> Status
move n Status{machine = m, widgets = w} = Status { machine = n
                                                 , widgets = w''
                                                 }
   where mWidget = readArray w m
         w'      = writeArray w  m (literal NoWidget)
         w''     = writeArray w' n mWidget

execute :: Status -> [Action] -> Status
execute initial actions = execState (run actions) initial
  where run :: [Action] -> State Status ()
        run = foldr (withState . move) (return ())

planN :: Int -> Predicate
planN ac = do actions <- mapM (\i -> free ("action" ++ show i)) [1 .. ac]

              let Status finalMachine finalWidgets = execute s0 actions

              return $ finalMachine .== literal MachineA &&& readArray finalWidgets finalMachine .== literal Widget5

   where s0 = Status { machine = literal MachineB
                     , widgets = mkSFunArray (\m -> ite (m .== literal MachineB) (literal Widget5) (literal NoWidget))
                     }

plan :: IO SatResult
plan = sat $ planN 1

main :: IO ()
main = print =<< plan
```

**Comment by LeventErkok** [comment]
Even with this richer `Action` type, I'd recommend following the structure I outlined, as opposed to trying to model `Maybe`. My personal experience is that trying to model sum-types almost always ends up being more confusing than the direct style. You shouldn't need any `Maybe`'s at all.

The issue in #343 is similar, and I have the same comments: Even if SBV supported SMT datatypes (which it doesn't aside from enumerations), the additional complexity and solver support makes it better to structure your code to avoid that path.

Of course, there's plenty of ways to solve the same problem; and what ends up best is what you're most comfortable with. I'm going to close this ticket as I don't think there's any action items on SBV for the time being; but feel free to open a new one if you have further design questions as it pertains to use of symbolic values.

**Comment by ocharles** [comment]
Thanks. What I have does at least work, but I'm going to study your suggestion some more and see if there's anything there that I can incorporate.

**Comment by LeventErkok** [comment]
The rule of thumb, I found, is to push symbolic values as deep down your data-structures as possible. Keep all your structures, complex data structures etc., concrete and at the Haskell level. (This is what the `Mergeable` class allows you to do, allowing you to treat them symbolically, without mapping them to the SMT solver.) I found this scales the best and takes advantage of what the SMT solver is good for, while leaving all the "infrastructure" work to Haskell, which it shines at.

Another similar example is the U2-bridge crossing one: http://hackage.haskell.org/package/sbv-7.9/docs/Documentation-SBV-Examples-Puzzles-U2Bridge.html

Another iterative deepening search example using queries is: http://hackage.haskell.org/package/sbv-7.9/docs/Documentation-SBV-Examples-Puzzles-HexPuzzle.html

Exactly what design would work is, of course, problem specific. But keeping symbolic values as deeply pushed down as possible is usually a good rule to follow.
